### PR TITLE
On-site potentials implemented

### PR DESCRIPTION
**Pull request by bhourahine**
On-site only potentials (preparation for SCC-MBD)
- [x] clean up electric field structures and routines
- [x] on site potential for hamiltonian and energy
- [x] on-site potentials in input
- [x] energy component printing
- [x] documentation
- [x] test examples
- [x] incompatible cases

Ehrenfest changes left to later PR, likewise use for MBD as next PR

### CODE REVIEW

File: prog/dftb+/lib_dftb/extfields.F90 [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+!                                                                                                  !
+!  See the LICENSE file for terms of usage and distribution.                                       !
+!--------------------------------------------------------------------------------------------------!
+
+!> External fields
+module dftbp_extfields
+  use dftbp_accuracy, only : dp, lc
+  use dftbp_potentials, only : TPotentials
+  use dftbp_periodic, only : TNeighbourList
+  use dftbp_message, only : error
```
Comment (awvwgk) [inline prog/dftb+/lib_dftb/extfields.F90:13]: Sort alphabetically
```suggestion
  use dftbp_message, only : error
  use dftbp_periodic, only : TNeighbourList
  use dftbp_potentials, only : TPotentials
```

File: prog/dftb+/lib_dftbplus/inputdata.F90 [hunk 2]
```diff
@@ -381,25 +386,11 @@ module dftbp_inputdata
     !> finite charge width if needed
     real(dp), allocatable :: extChrgBlurWidth(:)
 
+    !> Homogeneous external electric field
+    type(TElecField_inp), allocatable :: electricField
```
Comment (awvwgk) [inline prog/dftb+/lib_dftbplus/inputdata.F90:390]:
```suggestion
    type(TElecFieldInput), allocatable :: electricField
```

File: prog/dftb+/lib_dftbplus/inputdata.F90 [hunk 3]
```diff
@@ -381,25 +386,11 @@ module dftbp_inputdata
...
-    real(dp) :: EfieldVector(3)
-
-    !> frequency of time dependent field
-    real(dp) :: EfieldOmega
-
-    !> relative phase of field
-    integer :: EfieldPhase = 0
-
+    !> Potential(s) at atomic sites
+    type(TAtomExtPot_inp), allocatable :: atomicExtPotential
```
Comment (awvwgk) [inline prog/dftb+/lib_dftbplus/inputdata.F90:393]:
```suggestion
    type(TAtomExtPotInput), allocatable :: atomicExtPotential
```

File: prog/dftb+/lib_dftbplus/inputdata.F90 [hunk 4]
```diff
@@ -40,6 +40,8 @@ module dftbp_inputdata
 #:endif
   use dftbp_poisson, only : TPoissonInfo
   use dftbp_h5correction, only : TH5CorrectionInput
+  use dftbp_extfields, only : TElecField_inp
+  use dftbp_potentials, only : TAtomExtPot_inp
```
Comment (awvwgk) [inline prog/dftb+/lib_dftbplus/inputdata.F90:44]:
```suggestion
  use dftbp_extfields, only : TElecFieldInput
  use dftbp_potentials, only : TAtomExtPotInput
```

File: prog/dftb+/lib_dftb/potentials.F90 [hunk 5]
```diff
@@ -58,14 +60,39 @@ module dftbp_potentials
...
+
+    !> external atom and spin resolved onsite only potential (i.e. relating to net, not gross
+    !> populations)
+    real(dp), allocatable :: extOnSiteAtom(:,:)
+
   end type TPotentials
 
 
+  !> External potentials at atomic sites
+  type TAtomExtPot_inp
```
Comment (awvwgk) [inline prog/dftb+/lib_dftb/potentials.F90:75]:
```suggestion
  type TAtomExtPotInput
```

File: prog/dftb+/lib_dftb/potentials.F90 [hunk 6]
```diff
@@ -58,14 +60,39 @@ module dftbp_potentials
...
+    !> Atom(s) at which potential is present
+    integer, allocatable :: iAt(:)
+
+    !> Potentials experienced by net population charge on atom
+    real(dp), allocatable :: VextOnSite(:)
+
+    !> Atom(s) at which on-site potential is present
+    integer, allocatable :: iAtOnSite(:)
+
+  end type TAtomExtPot_inp
```
Comment (awvwgk) [inline prog/dftb+/lib_dftb/potentials.F90:89]:
```suggestion
  end type TAtomExtPotInput
```

File: prog/dftb+/lib_dftb/potentials.F90 [hunk 7]
```diff
@@ -79,6 +106,9 @@ subroutine TPotentials_init(this, orb, nAtom, nSpin)
     !> number of spins
     integer, intent(in) :: nSpin
 
+    !> Should the on site potentials be allocated?
+    type(TAtomExtPot_inp), intent(in), optional :: extAtPotentials
```
Comment (awvwgk) [inline prog/dftb+/lib_dftb/potentials.F90:110]:
```suggestion
    type(TAtomExtPotInput), intent(in), optional :: extAtPotentials
```

File: prog/dftb+/lib_dftb/potentials.F90 [hunk 8]
```diff
@@ -15,7 +15,7 @@ module dftbp_potentials
   implicit none
 
   private
-  public :: TPotentials, TPotentials_init
+  public :: TPotentials, TPotentials_init, TAtomExtPot_inp
```
Comment (awvwgk) [inline prog/dftb+/lib_dftb/potentials.F90:18]:
```suggestion
  public :: TPotentials, TPotentials_init, TAtomExtPotInput
```

File: prog/dftb+/lib_dftb/extfields.F90 [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+    !> is this an onsite (net charge related, so diagonal elements of hamiltonian) or not
+    !> (gross/Mulliken charge related, so incluing off diagonal coupling)
+    logical, allocatable :: atomicOnSite(:)
+
+  end type TEField
+
+
+  !> External homogeneous electric field
+  type TElecField_inp
```
Comment (awvwgk) [inline prog/dftb+/lib_dftb/extfields.F90:62]:
```suggestion
  type TElecFieldInput
```

File: prog/dftb+/lib_dftb/extfields.F90 [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    !> direction
+    real(dp) :: EfieldVector(3)
+
+    !> frequency of time dependent field
+    real(dp) :: EfieldOmega
+
+    !> relative phase of field
+    integer :: EfieldPhase = 0
+
+  end type TElecField_inp
```
Comment (awvwgk) [inline prog/dftb+/lib_dftb/extfields.F90:79]:
```suggestion
  end type TElecFieldInput
```

File: prog/dftb+/lib_dftb/extfields.F90 [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+!> External fields
+module dftbp_extfields
+  use dftbp_accuracy, only : dp, lc
+  use dftbp_potentials, only : TPotentials
+  use dftbp_periodic, only : TNeighbourList
+  use dftbp_message, only : error
+  implicit none
+
+  public :: TEField, TElecField_inp, setUpExternalField
```
Comment (awvwgk) [inline prog/dftb+/lib_dftb/extfields.F90:16]:
```suggestion
  public :: TEField, TElecFieldInput, setUpExternalField
```

File: prog/dftb+/lib_dftb/extfields.F90 [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    !> relative phase of field
+    integer :: EfieldPhase = 0
+
+  end type TElecField_inp
+
+contains
+
+  !> Sets up external electric or other fields
+  subroutine setUpExternalField(eField, tPeriodic, neighbourList, nNeighbourSK, iCellVec,&
+      & img2CentCell, cellVec, deltaT, iGeoStep, coord0Fold, coord, potential)
```
Comment (awvwgk) [inline prog/dftb+/lib_dftb/extfields.F90:85]: Maybe `addExternalField` is a better name?

**Review by awvwgk — APPROVED** [review]
Looks good to me, only some minor comments about naming of types and routines.

File: prog/dftb+/lib_dftb/extfields.F90 (hunk 12 above)
Comment (bhourahine) [reply prog/dftb+/lib_dftb/extfields.F90]: Done

**Pull request merged**